pkg/yqlib/encoder_toml.go

```diff
@@ -30,3 +65,596 @@ func (te *tomlEncoder) PrintLeadingContent(_ io.Writer, _ string) error {
 func (te *tomlEncoder) CanHandleAliases() bool {
 	return false
 }
+
+// ---- helpers ----
+
+func (te *tomlEncoder) writeComment(w io.Writer, comment string) error {
+	if comment == "" {
+		return nil
+	}
+	lines := strings.Split(comment, "\n")
+	for _, line := range lines {
+		line = strings.TrimSpace(line)
+		if !strings.HasPrefix(line, "#") {
+			line = "# " + line
+		}
+		if _, err := w.Write([]byte(line + "\n")); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (te *tomlEncoder) formatScalar(node *CandidateNode) string {
+	switch node.Tag {
+	case "!!str":
+		// Quote strings per TOML spec
+		return fmt.Sprintf("%q", node.Value)
+	case "!!bool", "!!int", "!!float":
+		return node.Value
+	case "!!null":
+		// TOML does not have null; encode as empty string
+		return "\"\""
+	default:
+		return node.Value
+	}
+}
+
+func (te *tomlEncoder) encodeRootMapping(w io.Writer, node *CandidateNode) error {
+	te.wroteRootAttr = false // Reset state
+
+	// Write root head comment if present (at the very beginning, no leading blank line)
+	if node.HeadComment != "" {
+		if err := te.writeComment(w, node.HeadComment); err != nil {
+			return err
+		}
+	}
+
+	// Preserve existing order by iterating Content
+	for i := 0; i < len(node.Content); i += 2 {
+		keyNode := node.Content[i]
+		valNode := node.Content[i+1]
+		if err := te.encodeTopLevelEntry(w, []string{keyNode.Value}, valNode); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// encodeTopLevelEntry encodes a key/value at the root, dispatching to attribute, table, or array-of-tables
+func (te *tomlEncoder) encodeTopLevelEntry(w io.Writer, path []string, node *CandidateNode) error {
+	switch node.Kind {
+	case ScalarNode:
+		// key = value
+		return te.writeAttribute(w, path[len(path)-1], node)
+	case SequenceNode:
+		// Empty arrays should be encoded as [] attributes
+		if len(node.Content) == 0 {
+			return te.writeArrayAttribute(w, path[len(path)-1], node)
+		}
+
+		// If all items are mappings => array of tables; else => array attribute
+		allMaps := true
+		for _, it := range node.Content {
+			if it.Kind != MappingNode {
+				allMaps = false
+				break
+			}
+		}
+		if allMaps {
+			key := path[len(path)-1]
+			for _, it := range node.Content {
+				// [[key]] then body
+				if _, err := w.Write([]byte("[[" + key + "]]\n")); err != nil {
```

Comment:
Here I'm playing the devil advocate role.

Are we sure the key cannot contain `[[` ? And so it would require some escaping.